stonesoup/dataassociator/tree.py

```diff
@@ -0,0 +1,176 @@
+    and at the same timestamp.
+    """
+    predictor = Property(
+        Predictor,
+        doc="Predict tracks to detection times")
+    updater = Property(
+        Updater,
+        doc="Updater used to get measurement prediction")
+    number_of_neighbours = Property(
+        int, default=None,
+        doc="Number of neighbours to find. Default `None`, which means all "
+            "points within the :attr:`max_distance` are returned.")
+    max_distance = Property(
+        float, default=np.inf,
+        doc="Max distance to return points. Default `inf`")
+
+    def generate_hypotheses(self, tracks, detections, time, **kwargs):
+        # No need for tree here.
+        if not tracks:
+            return set()
+        if not detections:
+            return {track: self.hypothesiser.hypothesise(
+                track, detections, time, **kwargs)
+                for track in tracks}
+
+        detections_list = list(detections)
+        tree = KDTree(
+            np.vstack([detection.state_vector[:, 0]
+                       for detection in detections_list]))
+
+        track_detections = defaultdict(set)
+        for track in tracks:
+            prediction = self.predictor.predict(track.state, time)
+            meas_pred = self.updater.predict_measurement(prediction)
+
+            if self.number_of_neighbours is None:
+                indexes = tree.query_ball_point(
+                    meas_pred.state_vector.ravel(),
+                    r=self.max_distance)
+            else:
+                _, indexes = tree.query(
+                    meas_pred.state_vector.ravel(),
+                    k=self.number_of_neighbours,
+                    distance_upper_bound=self.max_distance)
+
+            for index in np.atleast_1d(indexes):
+                if index != len(detections_list):
+                    track_detections[track].add(detections_list[index])
+
+        return {track: self.hypothesiser.hypothesise(
+            track, track_detections[track], time, **kwargs)
+            for track in tracks}
+
+
+class TPRTreeMixIn(DataAssociator):
+    measurement_model = Property(MeasurementModel)
+    horizon_time = Property(datetime.timedelta)
+    vel_mapping = Property(np.ndarray, default=None)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.vel_mapping is None:
+            self.pos_mapping = self.measurement_model.mapping[::2]
+            self.vel_mapping = self.measurement_model.mapping[1::2]
+        else:
+            self.pos_mapping = self.measurement_model.mapping
+
+        # Create tree
+        tree_property = rtree.index.Property(
+            type=rtree.index.RT_TPRTree,
+            tpr_horizon=self.horizon_time.total_seconds(),
+            dimension=len(self.pos_mapping))
+        self._tree = rtree.index.RtreeContainer(properties=tree_property)
+        self._coords = dict()
+
+    def _track_tree_coordinates(self, track):
```

Comment:
In cases where the measurement noise is much greater than the process/dynamic noise, and the time window is relatively small, the tracks seem to miss detections very often. 

One way of alleviating the above would be to consider the measurement noise covariance when computing `state_delta`, e.g.:
```python
state_delta = 3 * np.sqrt(
            np.diag(track.covar)[self.pos_mapping].reshape(-1, 1))
meas_delta = 3 * np.sqrt(np.diag(self.measurement_model.covar()).reshape(-1, 1))
state_delta += meas_delta
```

However, this may not necessarily be correct when it comes to individual detections, that carry their own measurement model.

Reply:
Currently the intersection query only uses the measurement as is, passing it's state vector in as min and max. This could be changed to use the measurement model error similar to track and this would then allow measurements with different models/errors.